### ISSUE TITLE
(Android) Unable to open PDFs after update

### ISSUE THREAD

**Issue by jonnyl2** [koreader/koreader#12098]
* KOReader version: 2024.04.209-g617618d58_2024-06-25
* Device: Huawei Mate 20 (hma-l29)
* OS: Android Q, api 29 on arm 
* Build flavor: rocks

#### Issue

After just updating OTA to the latest nightly, KOReader won't open any PDF files anymore and reports "No reader engine for this file or invalid file."

#### Steps to reproduce

Try to open any PDF/DJVU file with KOReader. 

<details>
  <summary>crash.log</summary>

```
--------- beginning of main
06-26 13:12:24.775 E/reader.launche( 9811): LoadAppImageStartupCache enabled : 1
06-26 13:12:24.781 E/reader.launche( 9811): Not starting debugger since process cannot load the jdwp agent.
06-26 13:12:24.816 E/BehaviorCollectManager( 9811): Fail to acquire dataAnalyzerService...
06-26 13:12:24.900 E/iGraphics( 9811): can not load local debug xml
06-26 13:13:30.977 E/reader.launche(10320): LoadAppImageStartupCache enabled : 1
06-26 13:13:30.981 E/reader.launche(10320): Not starting debugger since process cannot load the jdwp agent.
06-26 13:13:31.045 E/BehaviorCollectManager(10320): Fail to acquire dataAnalyzerService...
06-26 13:13:31.097 E/iGraphics(10320): can not load local debug xml
06-26 13:13:54.868 E/dlopen  (10320): error dlopen'ing /data/user/0/org.koreader.launcher/files/libs/libk2pdfopt.so.2: dlopen failed: cannot locate symbol "android_getCpuFamily" referenced by "/data/data/org.koreader.launcher/files/libs/libk2pdfopt.so.2"...
06-26 13:13:54.868 E/KOReader(10320): error: [string "dl.lua"]:73: error dlopen'ing /data/user/0/org.koreader.launcher/files/libs/libk2pdfopt.so.2: dlopen failed: cannot locate symbol "android_getCpuFamily" referenced by "/data/data/org.koreader.launcher/files/libs/libk2pdfopt.so.2"...
06-26 13:13:54.868 E/KOReader(10320): error: [string "dl.lua"]:167: could not find library /data/user/0/org.koreader.launcher/files/libs/libkoreader-djvu.so

```
</details>

Thank you.

**Comment by benoit-pierre** [comment]
Which looks suspiciously like the problem mentioned [here](https://github.com/koreader/koreader/discussions/12092#discussioncomment-9879020).

**Comment by pazos** [comment]
The APK update method updates files in place but doesn't remove old files unless they were replaced. This might introduce problems when shared libraries are updated, if they're versioned, because old ones are still there.

Just to double check this issue isn't caused by that you can either

a) uninstall the APK and reinstall the nightly directly.
b) go to KOReader app info -> storage -> wipe data and cache.

**Comment by benoit-pierre** [comment]
In this instance, that's not the problem. Fix incoming.

**Comment by benoit-pierre** [comment]
> The APK update method updates files in place but doesn't remove old files unless they were replaced. This might introduce problems when shared libraries are updated, if they're versioned, because old ones are still there.

Since those files are in the application private directory, could we not purge the directory before updating?

**Comment by Frenzie** [comment]
@benoit-pierre I don't have time right now to check what you mean, but that could be problematic. We always used to sign with a  debug key for convenience, but F-Droid needed it to be completely unsigned. But I think I kept the signing as part of kodev run android.

**Comment by pazos** [comment]
> Since those files are in the application private directory, could we not purge the directory before updating?

This should work:

```diff
diff --git a/app/src/main/java/org/koreader/launcher/Assets.kt b/app/src/main/java/org/koreader/launcher/Assets.kt
index a1ffe25..0307f3a 100644
--- a/app/src/main/java/org/koreader/launcher/Assets.kt
+++ b/app/src/main/java/org/koreader/launcher/Assets.kt
@@ -64,6 +64,7 @@ class Assets {

     private fun bootstrap(activity: Activity): Boolean {
         val filesDir = activity.filesDir.absolutePath
+        File(filesDir).deleteRecursively()
         activity.runOnUiThread { dialog = FramelessProgressDialog.show(activity, "") }

         /* copy regular files and extract 7z files from assets store */
```

**Comment by pazos** [comment]
IIRC debug builds are signed with AOSP debug keys. Release builds are unsigned.

**Comment by benoit-pierre** [comment]
My bad, I always build a release build with `./kodev release -i android`, but this time I had a leftover `KODEBUG=1` variable in the environment, to I ended up with a debug build.

**Comment by benoit-pierre** [comment]
> > Since those files are in the application private directory, could we not purge the directory before updating?
> 
> This should work:

After amending, to ensure the directory is created:
```diff
diff --git i/app/src/main/java/org/koreader/launcher/Assets.kt w/app/src/main/java/org/koreader/launcher/Assets.kt
index a1ffe25..e8664c7 100644
--- i/app/src/main/java/org/koreader/launcher/Assets.kt
+++ w/app/src/main/java/org/koreader/launcher/Assets.kt
@@ -64,6 +64,8 @@ class Assets {
 
     private fun bootstrap(activity: Activity): Boolean {
         val filesDir = activity.filesDir.absolutePath
+        File(filesDir).deleteRecursively()
+        File(filesDir).mkdir()
         activity.runOnUiThread { dialog = FramelessProgressDialog.show(activity, "") }
 
         /* copy regular files and extract 7z files from assets store */
```

**Comment by Frenzie** [comment]
I too have a bit of a hate hate relationship with Bionic.

**Comment by Frenzie** [comment]
Just a note that users of old Android readers in general tend to have a habit of doing things like this:
https://www.mobileread.com/forums/showthread.php?p=4433880#post4433880

**Comment by benoit-pierre** [comment]
With respect to nuking the application data files?

**Comment by Frenzie** [comment]
Yes. I don't know how much of a problem that is, but it's something to potentially be aware of.

**Comment by pazos** [comment]
> Yes. I don't know how much of a problem that is, but it's something to potentially be aware of.

Not a problem at all. We just wipe app data under `/data`. 

The only thing that *needs?* to stay in `/data` are hyphen dicts, which need to be pushed once each update, with a `1-userpatch.lua`. That will work as is.

External path `/sdcard/koreader` is used for everything else, AFAIK.

**Comment by Frenzie** [comment]
I know, I said use /sdcard and it came out as /data but that happens frequently. But I suppose the same people should know not to update if they do that.

**Comment by pazos** [comment]
Yup, `/data/media/0` -eq `/sdcard` on some devices.

**Comment by mergen3107** [comment]
Has the original issue been fixed?
I am hesitant to update my Onyx Boox from old October release :D

**Comment by Frenzie** [comment]
Assuming base was bumped.

**Comment by mergen3107** [comment]
Thanks @Frenzie 
It seems 1841 from base hasn’t made it to here yet.

**Comment by benoit-pierre** [comment]
Issues in this repo fixed in base are closed once a base bump containing the fix is done, no?

So there's no need to ask: wait for the issue to be closed + 24H for the fix to make it to the nightlies.

**Comment by pazos** [comment]
@jonnyl2: wipping application data should fix your issue. Uninstalling and reinstalling is just a hardcore way to wipe application data :)

**Comment by jonnyl2** [comment]
@pazos Did you mean clearing the application data & cache in the Android Apps settings? I tried that, but it didn't work. I then uninstalled the app and reinstalled a freshly downloaded nightly APK, and it is working now!

**Comment by pazos** [comment]
> @pazos Did you mean clearing the application data & cache in the Android Apps settings? I tried that, but it didn't work. I then uninstalled the app and reinstalled a freshly downloaded nightly APK, and it is working now!

Thanks for letting me know. One thing less to recommend :)

Both do the same, but maybe some devices are picky about what to prune when the app is running. So force-closing first might be mandatory.

But uninstall - install is way easier :)

**Comment by benoit-pierre** [comment]
Fix merged in #12140.
